src/storage/src/source/postgres/snapshot.rs

```diff
@@ -0,0 +1,376 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Renders the table snapshot side of the [`PostgresSourceConnection`] ingestion dataflow.
+//!
+//! # Snapshot reading
+//!
+//! Depending on the resumption LSNs the table reader decides which tables need to be snapshotted
+//! and performs a simple `COPY` query on them in order to get a snapshot. There are a few subtle
+//! points about this operation, described in the following sections.
+//!
+//! ## Consistent LSN point for snapshot transactions
+//!
+//! Given that all our ingestion is based on correctly timestamping updates with the LSN they
+//! happened at it is important that we run the `COPY` query at a specific LSN point that is
+//! relatable with the LSN numbers we receive from the replication stream. Such point does not
+//! necessarily exist for a normal SQL transaction. To achieve this we must force postgres to
+//! produce a consistent point and let us know of the LSN number of that by creating a replication
+//! slot as the first statement in a transaction.
+//!
+//! This is a temporary dummy slot that is only used to put our snapshot transaction on a
+//! consistent LSN point. Unfortunately no lighterweight method exists for doing this. See this
+//! [postgres thread] for more details.
```

Comment:
We should mention why we don't just create a real replication slot and use that in the `replication` module, instead opting for this temporary one! Also, thank you for linking to the email thread!